tidy3d/plugins/autograd/functions.py

```
@@ -180,19 +313,27 @@ def convolve(
     if any(k % 2 == 0 for k in kernel.shape):
         raise ValueError(f"All kernel dimensions must be odd, got {kernel.shape}.")
 
-    if kernel.ndim != array.ndim and axes is None:
-        raise ValueError(
-            f"Kernel dimensions must match array dimensions, got kernel {kernel.shape} and array {array.shape}."
-        )
-
-    if mode in ("same", "full"):
-        kernel_dims = kernel.shape if axes is None else [kernel.shape[d] for d in axes[1]]
-        pad_widths = [(ks // 2, ks // 2) for ks in kernel_dims]
-        for axis, pad_width in enumerate(pad_widths):
-            array = pad(array, pad_width, mode=padding, axis=axis)
-        mode = "valid" if mode == "same" else mode
-
-    return convolve_ag(array, kernel, axes=axes, mode=mode)
+    axes_array, axes_kernel = _normalize_axes(array.ndim, kernel.ndim, axes)
+
+    working_array = array
+    effective_mode = mode
+
+    if mode in ["same", "full"]:
+        for ax_array, ax_kernel in zip(axes_array, axes_kernel):
+            if mode == "same" and kernel.shape[ax_kernel] % 2 == 0:
+                raise ValueError(
+                    f"Even-sized kernel along axis {ax_kernel} not supported for 'same' mode."
+                )
```

Does numpy also complain about this?

---

no, it silently returns a result that is shifted by half a sample (floor division), i.e. it assumes center = floor(M/2) for kernel length M - which I dislike... But guess we want to be compatible with that?

---

Hmm yeah I think it's probably fine as-is, we're not really trying to match the scipy or numpy interface here